49_group-anagrams.md

```diff
@@ -0,0 +1,144 @@
+# 49. Group Anagrams
+
+## 1st
+
+### ①
+
+各文字列をsortすることでcanonicalな形にしてそれをkeyにしてグループ分けする。
+`0 <= strs[i].length <= 100` の制約から、Cで実装されたsorted() (O(nlogn))の方がO(n)なPythonの関数よりも速いだろうと考えそちらを使った。
+    - https://github.com/niklas-heer/speed-comparison
+    - https://benchmarksgame-team.pages.debian.net/benchmarksgame/box-plot-summary-charts.html
+
+
+所要時間: 4:57
+
+n: len(strs), m: str_の長さの平均
+- 時間計算量: O(nm*log(m))
+- 空間計算量: O(nm)
+
+```py
+class Solution:
+    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
+        sorted_str_to_anagrams = defaultdict(list)
+        for str_ in strs:
```

Comment:
普通の変数名でtrailing underscoreは一般的ですか？

Reply:
https://peps.python.org/pep-0008/#descriptive-naming-styles

> In addition, the following special forms using leading or trailing underscores are recognized (these can generally be combined with any case convention):
> (省略)
> single_trailing_underscore_: used by convention to avoid conflicts with Python keyword, e.g. :
> ```py
> tkinter.Toplevel(master, class_='ClassName')
> ```

なので、keywordとの衝突を避けるときに使うのは一般的だという認識です。ただstrはkeywordではなくbuilt-in functionなので正確には違うんですが、そこはいいんじゃないかな？と解釈しています

Reply:
私は、これくらい明らかならば、一文字変数でも別にいいんじゃないかなと個人的には思います。

一文字は、文脈からなにかがはっきり分かり、使ったらすぐに忘れていい、という意味です。

Reply:
> 私は、これくらい明らかならば、一文字変数でも別にいいんじゃないかな

自分もそう思います。今回は単数形と複数形の対比がはっきりしていると思ったのでこちらを選びました。自分が考えるよりも他の方のtrailing underscoreへの抵抗感は強そうだと感じました...

https://peps.python.org/pep-0008/#function-and-method-arguments
> Perhaps better is to avoid such clashes by using a synonym.

とあるし、もう少し別の語を使って衝突を避ける方向で行こうと思います